Rework tokio_tcp and tokio_udp re-exports in tokio::net

casimir opened the pull request:
This is the continuation of #526. I applied the same pattern than before.

To signify deprecation I had to use type aliases since it not possible to deprecate a re-export for now (see https://github.com/rust-lang/rust/issues/30827 for details).

I also split the documentation in the different modules. Documentation wording isn't really my forte so feel free to suggest enhancements on this side.

Last thing I made the `unix` module present but empty on non-unix systems. I'm not sure this is the best idea but it allows completeness on all systems.

carllerche commented on src/net.rs:
```diff
@@ -1,54 +1,71 @@
+//! TCP/UDP/Unix bindings for `tokio`.
 //!
-//! This module contains the TCP/UDP networking types, similar to the standard
+//! This module contains the TCP/UDP/Unix networking types, similar to the standard
```
I agree that the current documentation should be moved into the appropriate sub module, but that leaves `tokio::net` documentation a bit light, especially given that it is a primary module.

Perhaps, you could add an "orgnaization" section, similar to https://doc.rust-lang.org/std/net/index.html. 

Also, this can mention the unix module and explain that it is only available on unix platforms.

hawkw replied:
Looks like this change was made in 7ea54a01ffee8b06cbfaf887bf15daf9aa696f10.

carllerche commented on src/net.rs:
```diff
@@ -1,54 +1,71 @@
+pub use self::tcp::{TcpListener, TcpStream};
+
+#[deprecated(note = "use `tokio::net::tcp::ConnectFuture` instead")]
+/// Convenience re-export of [`tcp::ConnectFuture`](tcp/struct.ConnectFuture.html). Will be removed in a future release.
```
I would just remove docs on deprecated items and also include a `#[doc(hidden)]` annotation.

carllerche commented on src/net.rs:
```diff
@@ -1,54 +1,71 @@
+pub type SendDgram<T> = self::udp::SendDgram<T>;
 
-#[cfg(unix)]
 pub mod unix {
```
I'm a bit torn about this, but I would suggest we stick to the current idiom of having the cfg on the entire module vs. the types inside.

I think we can achieve the same level of documentation by having a `# Unix` section in `tokio::net` explaining how it works.

hawkw replied:
Again, looks like this change was made in 7ea54a01ffee8b06cbfaf887bf15daf9aa696f10.

carllerche commented:
Any updates?

casimir commented:
Actually yes, just waited for some time to submit the changes. I think I addressed all your comments.

Also, I went to great lengths to add [platform specific documentation](https://doc.rust-lang.org/beta/rustdoc/unstable-features.html#documenting-platform-feature-specific-information) to get a result similar to [`std::os::*`](https://doc.rust-lang.org/std/os/unix/index.html). Sadly the result was underwhelming, it doesn't work well with crate reexport (it most just an empty `unix` module on windows...) so I finally deleted it.
